### PR TITLE
CORE-16615: introduce new api to find an unconsumed states

### PR DESCRIPTION
**Pull request by jennyang-r3**
## Overview
This PR introduces a new API that finds unconsumed states given the state type. We are introducing this function even though there is `findUnconsumedStatesByType` because the function uses in-memory filtering, whereas this new function uses db query index.

### CODE REVIEW

File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/UtxoLedgerService.java [hunk 1]
```diff
@@ -87,7 +87,8 @@ public interface UtxoLedgerService {
     UtxoFilteredTransactionBuilder filterSignedTransaction(@NotNull UtxoSignedTransaction transaction);
 
     /**
-     * Finds unconsumed states of the specified {@link ContractState} type in the vault.
+     * Finds unconsumed states and subclasses of the states of the specified {@link ContractState} type in the vault.
+     * Only use this if you really care about catching all child classes, it has poor performance.
```
Comment (vlajos) [inline ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/UtxoLedgerService.java]:
```suggestion
     * Finds unconsumed states of the specified {@link ContractState} type and its subclasses in the vault.
     * Only use this if you really care about catching all child classes. {@link findUnconsumedStatesByExactType} performs better for exact-matching use cases.
```
Sounds a bit more positive to me.

**Review by vlajos — DISMISSED** [review]

**Review by vlajos — DISMISSED** [review]

**Review by relyafi — DISMISSED** [review]
LGTM

File: data/avro-schema/src/main/resources/avro/net/corda/data/ledger/persistence/FindUnconsumedVisibleExactStates.avsc [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "type": "record",
+  "name": "FindUnconsumedStatesByExactType",
+  "doc": "Retrieve the unconsumed visible states of specific type. This only retrieves exact type excluding its states of subclass unlike {@link FindUnconsumedStatesByExactType}. One of several types of ledger persistence request {@link LedgerPersistenceRequest}",
```
Comment (blsemo) [inline data/avro-schema/src/main/resources/avro/net/corda/data/ledger/persistence/FindUnconsumedVisibleExactStates.avsc]: I think you are comparing with the old `FindUnconsumedStatesByType` record.
```suggestion
  "doc": "Retrieve the unconsumed visible states of specific type. This only retrieves exact type excluding its states of subclass unlike {@link FindUnconsumedStatesByType}. One of several types of ledger persistence request {@link LedgerPersistenceRequest}",
```

**Review by blsemo — COMMENTED** [review]
I think there's an error in the AVRO documentation.

File: data/avro-schema/src/test/kotlin/net/corda/data/ledger/persistence/LedgerPersistenceRequestSchemaCompatibilityTest.kt [hunk 3]
```diff
@@ -50,6 +50,7 @@ class LedgerPersistenceRequestSchemaCompatibilityTest {
                 "net.corda.data.ledger.persistence.PersistTransactionIfDoesNotExist",
                 "net.corda.data.ledger.persistence.FindTransaction",
                 "net.corda.data.ledger.persistence.FindUnconsumedStatesByType",
+                "net.corda.data.ledger.persistence.FindUnconsumedStatesByExactType",
```
Comment (lankydan) [inline data/avro-schema/src/test/kotlin/net/corda/data/ledger/persistence/LedgerPersistenceRequestSchemaCompatibilityTest.kt]: I don't think we want changes to this file.

@vlajos I think you said we won't support 5.0 to 5.1 upgrades, therefore we _could_ make these changes in preparation of 5.1 to 5.2 upgrades. But, solely for the changes made in this PR, we don't need these.

File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/UtxoLedgerService.java [hunk 4]
```diff
@@ -87,7 +87,8 @@ public interface UtxoLedgerService {
     UtxoFilteredTransactionBuilder filterSignedTransaction(@NotNull UtxoSignedTransaction transaction);
 
     /**
-     * Finds unconsumed states of the specified {@link ContractState} type in the vault.
+     * Finds unconsumed states and subclasses of the states of the specified {@link ContractState} type and its subclasses in the vault.
+     * Only use this if you really care about catching all child classes. {@link #findUnconsumedStatesByExactType(Class<T>)} performs better for exact-matching use cases.
```
Comment (lankydan) [inline ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/UtxoLedgerService.java]:
```
* Finds unconsumed states and subclasses of the states of the specified {@link ContractState} type and its subclasses in the vault.
* Only use this if you really care about catching all child classes. {@link #findUnconsumedStatesByExactType(Class<T>)} performs better for exact-matching use 
```

to

```
* Finds unconsumed states that are concrete implementations or subclasses of {@code type}. 
* <p>
* Only use this method if subclasses of {@code type} must be returned.
* <p>
* Use {@link #findUnconsumedStatesByExactType(Class<T>)} to return exact instances of the input {@code type}. This method is more performant than {@link #findUnconsumedStatesByType(Class<T>)}.
* <p>
* Use {@link #query(String, Class)} for a more performant method of retrieving subclasses of a specified type.
```

What do you think?

File: data/avro-schema/src/test/kotlin/net/corda/data/ledger/persistence/LedgerPersistenceRequestSchemaCompatibilityTest.kt (hunk 3 above)
Comment (vlajos) [reply data/avro-schema/src/test/kotlin/net/corda/data/ledger/persistence/LedgerPersistenceRequestSchemaCompatibilityTest.kt]: AFAIK only 5.0-5.1 flow communication is not a requirement at the moment, but still, I think 
you are right.
`oldSchemaJson` should have the 5.0 schema. The new API is not a 5.0 one.

**Review by lankydan — APPROVED** [review]

**Review by blsemo — APPROVED** [review]
LGTM

**Pull request merged**